### ISSUE TITLE
quantile_rank

### ISSUE THREAD

**Issue by jonlachlan** [simple-statistics/simple-statistics#119]
I'm looking for a way to calculate the quantile_rank of a specific data point. For example, I may want to know that a data point is in the 80th percentile, or in the 3rd quartile. I think I'd prefer a precise value, for example "this value is at the 3.245 quartile", which I can round to 3 if I want.

I've done this in Postgres using the ntile() window function (http://www.postgresql.org/docs/9.4/static/functions-window.html), but I'd like to do it in Javascript.

**Comment by jonlachlan** [comment]
I suppose if I have a sorted array, I know the rank of each value, then take that as a percentage of the array length, then multiply by the number of quantiles I'm are looking for.

So if 75 is the 100th element in an array of 200, then this is the 100/200 or .5. Multiply by 100 to get the percentile rank (50) or multiply by 4 for the quartile rank (2).

I think I'm good do this on my own, but thanks anyways :)

**Comment by tmcw** [comment]
so the way to do this would probably be something like

```js
var myNumber = ...;
var myArray = [...];
var qRank = ss.bisect(myArray.sort(), myNumber) / myArray.length;
```

**Comment by jonlachlan** [comment]
Hmm I'm not seeing a bisect function, is that in a new ss dist?

**Comment by tmcw** [comment]
There currently isn't one - I'm proposing we could include one

**Comment by denizdogan** [comment]
Has anyone made any progress on this? I have an unsorted array of values as `[x, y, z, w]` and I'm looking to know what percentile `z` is in, is that a separate issue or is this the one? I noticed that [stats-lite](https://www.npmjs.com/package/stats-lite) has such a function, but since I already use simple-statistics, I'm not too keen on switching now.

**Comment by Yomguithereal** [comment]
@denizdogan it seems the `bisect` function is now part of the current release on npm. Have you tried @tmcw's solution?

**Comment by Yomguithereal** [comment]
What's more, I don't see a function such as the one you need in the `stats-lite` module.

**Comment by Yomguithereal** [comment]
@tmcw on a side note, should we implement a quantile_rank function based on scipy? https://github.com/scipy/scipy/blob/v1.0.0/scipy/stats/stats.py#L1709-L1802

**Comment by denizdogan** [comment]
@Yomguithereal I spoke too quickly, the function in stats-lite is slightly different than the one I proposed. Anyways, I honestly don't understand how I'm supposed to use the `bisect` function, its signature is different than the one in the comment above. :/

**Comment by Yomguithereal** [comment]
@denizdogan yes it seems the signature changed to take a function rather than an array. So instead, you should probably use a basic binary search (I assume the value you want the percentile from is in your array, else it would be slightly more complex:

```js
function binarySearch(array, value) {
  var mid = 0;
  var lo = 0;
  var hi = array.length;

  hi--;

  var current;

  while (lo <= hi) {
    mid = (lo + hi) >>> 1;

    current = array[mid];

    if (current > value) {
      hi = ~-mid;
    }
    else if (current < value) {
      lo = -~mid;
    }
    else {
      return mid;
    }
  }

  return -1;
}
```

Then:

```js
var qRank = binarySearch(myArray.sort(), myNumber) / myArray.length;
```

**Comment by Yomguithereal** [comment]
To get the percentile, multiply by 100 obviously :)

**Comment by Yomguithereal** [comment]
Or, even simpler if you don't care about performance:

```js
var qRank = myArray.sort().indexOf(myNumber) / myArray.length;
```

**Comment by denizdogan** [comment]
@Yomguithereal Thanks a lot, much appreciated! :)

**Comment by Yomguithereal** [comment]
Note however that if you have the found value multiple times, this formula is not completely correct and there are many ways to answer the question. For instance, default scipy would get the mean of the indices where your value is found in the sorted array.

**Comment by MohammedAlkindi** [comment]
Closing the loop on an old one: `quantileRank` / `quantileRankSorted` shipped in #285 (merged 2018-03-23, drafted from this very thread), so the ask here has been in the library for years. Verified at current `main` (7063073):

```js
import * as ss from "simple-statistics";
ss.quantileRank([1, 2, 4, 5, 6, 7, 8, 9], 4); // 0.375
typeof ss.quantileRankSorted;                 // "function"
```

Looks closeable as completed by #285.
